docs/reference/troubleshooting/corruption-issues.asciidoc

```diff
@@ -0,0 +1,92 @@
+[[corruption-troubleshooting]]
+== Troubleshooting corruption
+
+{es} expects that the data it reads from disk is exactly the data it previously
+wrote. If it detects that the data on disk is different from what it wrote then
+it will report some kind of exception such as:
+
+- `org.apache.lucene.index.CorruptIndexException`
+- `org.elasticsearch.gateway.CorruptStateException`
+- `org.elasticsearch.index.translog.TranslogCorruptedException`
+
+Typically these exceptions happen due to a checksum mismatch. Most of the data
+that {es} writes to disk is followed by a checksum using a simple algorithm
+known as CRC32 which is fast to compute and good at detecting the kinds of
+random corruption that may happen when using faulty storage. A CRC32 checksum
+mismatch definitely indicates that something is faulty, although of course a
+matching checksum doesn't prove the absence of corruption.
+
+Verifying a checksum is expensive since it involves reading every byte of the
+file which takes significant effort and might evict more useful data from the
+filesystem cache, so systems typically doesn't verify the checksum on a file
+very often. This is why you tend only to encounter a corruption exception when
+something unusual is happening. For instance, corruptions are often detected
+during merges, shard movements, and snapshots. This does not mean that these
+proceses are causing corruption: they are examples of the rare times where
+reading a whole file is necessary. {es} takes the opportunity to verify the
+checksum at the same time, and this is when the corruption is detected and
+reported. It doesn't indicate the cause of the corruption or when it happened.
+Corruptions can remain undetected for many months.
+
+The files that make up a Lucene index are written sequentially from start to
+end and then never modified or overwritten. This access pattern means the
+checksum computation is very simple and can happen on-the-fly as the file is
+initially written, and also makes it very unlikely that an incorrect checksum
+is due to a userspace bug at the time the file was written. The routine that
+computes the checksum is straightforward, widely used, and very well-tested, so
+you can be very confident that a checksum mismatch really does indicate that
+the data read from disk is different from the data that {es} previously wrote.
+
+The files that make up a Lucene index are written in full before they are used.
+If a file is needed to recover an index after a restart then your storage
+system will previously have confirmed to {es} that this file was durably synced
```

Comment:
Did you mean `had previously confirmed` here? `will previously have confirmed` sounds a bit strange to me.

Reply:
This one is legitimate English - if you want the technical details, `will have confirmed` is the future perfect tense, and that's acceptable to use in the first conditional form as I'm doing here. But I think it's fair to say that it could be confusing too, so I simplified it in 85a8ee2.